File: examples/python/inventory/handlers/reserve_stock.py
"""Handler for ReserveStock command."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import types_pb2 as types
from errors import CommandRejectedError
from proto import domains_pb2 as domains

from handlers.state import InventoryState


def handle_reserve_stock(
    command_book: types.CommandBook,
    command_any: Any,
    state: InventoryState,
    seq: int,
) -> types.EventBook:
    """Handle ReserveStock command."""
    if not state.exists():
        raise CommandRejectedError("Inventory not initialized")

    cmd = domains.ReserveStock()
    command_any.Unpack(cmd)

    if cmd.quantity <= 0:
        raise CommandRejectedError("Quantity must be positive")
    if not cmd.order_id:
        raise CommandRejectedError("Order ID is required")
    if cmd.order_id in state.reservations:
        raise CommandRejectedError("Reservation already exists for this order")
    if state.available() < cmd.quantity:
        raise CommandRejectedError(
            f"Insufficient stock: available {state.available()}, requested {cmd.quantity}"
        )

    new_available = state.available() - cmd.quantity

    pages = []

    reserved_event = domains.StockReserved(
        quantity=cmd.quantity,
        order_id=cmd.order_id,
        new_available=new_available,
        reserved_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )
    reserved_any = Any()
    reserved_any.Pack(reserved_event, type_url_prefix="type.examples/")
    pages.append(
        types.EventPage(
            num=seq,
            event=reserved_any,
            created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
        )
    )

    if new_available < state.low_stock_threshold and state.available() >= state.low_stock_threshold:
        alert_event = domains.LowStockAlert(
            product_id=state.product_id,
            available=new_available,
            threshold=state.low_stock_threshold,
            alerted_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
        )
        alert_any = Any()
        alert_any.Pack(alert_event, type_url_prefix="type.examples/")
        pages.append(
            types.EventPage(
                num=seq + 1,
                event=alert_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        )

    return types.EventBook(cover=command_book.cover, pages=pages)

File: examples/go/inventory/logic/reserve_stock.go
package logic

import (
	"inventory/proto/examples"

	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/timestamppb"
)

func (l *DefaultInventoryLogic) HandleReserveStock(state *InventoryState, quantity int32, orderID string) ([]proto.Message, error) {
	if !state.Exists() {
		return nil, NewFailedPrecondition("Inventory not initialized")
	}
	if quantity <= 0 {
		return nil, NewInvalidArgument("Quantity must be positive")
	}
	if orderID == "" {
		return nil, NewInvalidArgument("Order ID is required")
	}
	if _, exists := state.Reservations[orderID]; exists {
		return nil, NewFailedPrecondition("Reservation already exists for this order")
	}
	if state.Available() < quantity {
		return nil, NewFailedPreconditionf("Insufficient stock: available %d, requested %d", state.Available(), quantity)
	}

	events := []proto.Message{
		&examples.StockReserved{
			Quantity:     quantity,
			OrderId:      orderID,
			NewAvailable: state.Available() - quantity,
			ReservedAt:   timestamppb.Now(),
		},
	}

	// Check if we should trigger low stock alert
	newAvailable := state.Available() - quantity
	if newAvailable < state.LowStockThreshold && state.Available() >= state.LowStockThreshold {
		events = append(events, &examples.LowStockAlert{
			ProductId: state.ProductID,
			Available: newAvailable,
			Threshold: state.LowStockThreshold,
			AlertedAt: timestamppb.Now(),
		})
	}

	return events, nil
}
